qa/workunits/rbd/compare_mirror_images.sh

```diff
@@ -0,0 +1,193 @@
+             | jq 'select(.name | contains("mirror.primary"))' \
+             | jq 'select(.namespace.state == "demoted")')
+  if [[ $RBD_DEVICE_TYPE == "nbd" ]]; then
+    demote_id=$(echo $demote | jq -r '.id')
+    bdev=$(sudo rbd --cluster ${CLUSTER1} device map -t ${RBD_DEVICE_TYPE} \
+             -o try-netlink --snap-id ${demote_id} ${POOL}/${img})
+  elif [[ $RBD_DEVICE_TYPE == "krbd" ]]; then
+    demote_name=$(echo $demote | jq -r '.name')
+    bdev=$(sudo rbd --cluster ${CLUSTER1} device map -t ${RBD_DEVICE_TYPE} \
+             ${POOL}/${img}@${demote_name})
+  fi
+  demote_md5=$(sudo md5sum ${bdev} | awk '{print $1}')
+  sudo rbd --cluster ${CLUSTER1} device unmap -t ${RBD_DEVICE_TYPE} ${bdev}
+
+  wait_for_non_primary_demoted_mirror_snap ${CLUSTER2} ${POOL} ${img}
+
+  # promote non-primary image and calculate hash of its latest mirror snapshot
+  promote_image ${CLUSTER2} ${POOL} ${img}
+  promote=$(rbd --cluster ${CLUSTER2} snap ls --all ${POOL}/${img} \
+              --format=json \
+              | jq 'last' \
+              | jq 'select(.name | contains("mirror.primary"))')
+  if [[ $RBD_DEVICE_TYPE == "nbd" ]]; then
+    promote_id=$(echo $promote | jq -r '.id')
+    bdev=$(sudo rbd --cluster ${CLUSTER2} device map -t ${RBD_DEVICE_TYPE} \
+             -o try-netlink --snap-id ${promote_id} ${POOL}/${img})
+  elif [[ $RBD_DEVICE_TYPE == "krbd" ]]; then
+    promote_name=$(echo $promote | jq -r '.name')
+    bdev=$(sudo rbd --cluster ${CLUSTER2} device map -t ${RBD_DEVICE_TYPE} \
+             ${POOL}/${img}@${promote_name})
+  fi
+  promote_md5=$(sudo md5sum ${bdev} | awk '{print $1}')
+  sudo rbd --cluster ${CLUSTER2} device unmap -t ${RBD_DEVICE_TYPE} ${bdev}
+
+  if [ "${demote_md5}" != "${promote_md5}" ]; then
+    echo "demote_md5:${demote_md5} != promote_md5:${promote_md5} for \
+      pool/img:${POOL}/${img}"
+    return 1
+  fi
+}
+
+setup
+
+start_mirrors ${CLUSTER1}
+start_mirrors ${CLUSTER2}
+
+for i in {1..10}; do
+  for j in {1..10}; do
+    IMG=${IMG_PREFIX}${j}
+    MNTPT=${MNTPT_PREFIX}${j}
+    create_image_and_enable_mirror ${CLUSTER1} ${POOL} ${IMG} \
+      ${RBD_MIRROR_MODE} 10G
+    if [[ $RBD_DEVICE_TYPE == "nbd" ]]; then
+      BDEV=$(sudo rbd --cluster ${CLUSTER1} device map -t ${RBD_DEVICE_TYPE} \
+               -o try-netlink ${POOL}/${IMG})
+    elif [[ $RBD_DEVICE_TYPE == "krbd" ]]; then
+      BDEV=$(sudo rbd --cluster ${CLUSTER1} device map -t ${RBD_DEVICE_TYPE} \
+               ${POOL}/${IMG})
+    else
+      echo "Unknown RBD_DEVICE_TYPE: ${RBD_DEVICE_TYPE}"
+      return 1
+    fi
+    sudo mkfs.ext4 ${BDEV}
+    sudo mkdir -p ${MNTPT}
+    sudo mount ${BDEV} ${MNTPT}
+    # create mirror snapshots under I/O
+    launch_manual_msnaps ${CLUSTER1} ${POOL} ${IMG} &
+    run_bench ${MNTPT} ${WORKLOAD_TIMEOUT} &
+  done
+  wait
+
+  pids=''
+  for j in {1..10}; do
+    compare_demoted_promoted_mirror_snaps $j &
+    pids+=" $!"
+  done
+
+  for pid in $pids; do
+    wait "$pid"
```

Comment:
Maybe just a `wait` that'd wait for the completion for all the background processes would be sufficient.?

Reply:
Just `wait` would ignore exit codes for those background processes:
> Waits for each process identified by an ID, which may be a process ID or a job specification, and reports its termination status. If ID is not given, waits for all currently active child processes, and the return status is zero.

... so, if there is a checksum mismatch, it would make it to the output but wouldn't cause the job to fail.

Reply:
:man_facepalming:

Questions on your original comment,  https://github.com/ceph/ceph/pull/54802#discussion_r1466703267
I think it's okay for the script to terminate when one of the background compare job fails.  So,
``` 
for pid in $pids; do
    wait "$pid"
done
```
should be good enough?

Is there a need to wait for all the background compare jobs, and then terminate if at least one of them failed? In that case, we'd have to `set +e` temporarily in the script to collect the exit statuses of all the jobs.

Reply:
> Is there a need to wait for all the background compare jobs, and then terminate if at least one of them failed?

It would be nice, because you could terminate with a bunch of processes still running on the background otherwise.  Not a big of deal for teuthology, but could be an annoyance when running locally.

> In that case, we'd have to `set +e` temporarily in the script to collect the exit statuses of all the jobs.

Not necessarily.  Something like:
```
ret=0
for pid in $pids; do
    wait "$pid" || ret=$?
done
if ((ret != 0)); then
  echo "compare_demoted_promoted_images failed"
  exit 1
fi
```

Reply:
Thanks for the suggestion. Done.